### ISSUE TITLE
xm.optimizer_step does not state that it takes the average gradient across cores

### ISSUE THREAD

**Issue by hyang0129** [pytorch/xla#2840]
torch_xla.core.xla_model.optimizer_step calls the torch_xla.core.xla_model.reduce_gradients method, which implicitly executes all_reduce(REDUCE_SUM, gradients, scale=1.0 / count, groups=groups, cctx=cctx). 

Most users (me) would expect the optimizer to make a step of size LR, but it actually makes a step of size LR/cctx.world_size. This is opposite to the expected behaviour when converting a cuda single device GPU operation to an xla 8 core TPU operation. As in, it is reasonable for users to expect LR = x to actually mean LR = x and not LR = x/cores. Since the code does not behave as expected, it should be documented. 

Please add a note to the xm.optimizer_step to indicate that gradients used in this optimizer step are actually the average gradient across all cores and not the sum of gradients across all cores.

**Comment by taylanbil** [comment]
This depends on how your loss function is reduced. For example, take `nn.CrossEntropyLoss`, as in [this example](https://github.com/pytorch/xla/blob/08ae1044c2/test/test_train_mp_imagenet.py#L199).

If you are using loss reduction `sum`, then you would expect the gradients to be added up. If you are using loss reduction `mean`, then the gradients should be averaged instead. Since the default reduction for `nn.CrossEntropyLoss` and most other losses are using `mean`, it seems sensible to divide by the world_size when gradients are all-reduced. Am I missing anything?

For more complex losses, where it makes sense to normalize using, say, the number of tokens processed in a batch (which may vary from core to core), I'd recommend to skip using `xm.optimizer_step` and use `xm.reduce_gradients` + `xm.mesh_reduce` the denominator and normalize manually.

**Comment by hyang0129** [comment]
Yes that is what I would expect. However, comparing same batch size and learning rate using nn.CrossEntropyLoss and a SGD optimizer between GPU and TPU results in the TPU making significantly smaller gradient updates. Would it be helpful if I implement an example in google colab to reproduce this behaviour? 

Essentially, one would reasonably expect that nn.CrossEntropyLoss using mean reduction would result in a similar magnitude of gradient updates if we are using the same batch size in both GPU and TPU scenarios.

**Comment by taylanbil** [comment]
Example would be nice, but before that, can you describe the two setups? how many gpus/tpus, what is the **per device** batch size, etc?

**Comment by hyang0129** [comment]
1 GPU vs 1 TPU with 8 cores. Per device batch size is 16 for TPU and 128 for GPU. The total batch size should be 128 for both

**Comment by taylanbil** [comment]
and what kind of loss are you using? Can you include the training loop code?

**Comment by hyang0129** [comment]
nn.CrossEntropy with SGD. I'll reproduce the issue

**Comment by hyang0129** [comment]
The real issue lies in not reading the documentation for the Parallel loader. Both nn.DataParallel and Tensorflow 2.0 shard an existing batch across the GPUs or TPU cores. Parallel loader allocates a batch to a core, similar to how a dealer would deal cards. 

This behaviour isn't very clear until you read the argument devices : The list of devices where the data has to be sent. The i-th sample returned by the loader will be sent to devices[i % len(devices)]. Perhaps expanding the docstring from "Wraps an existing PyTorch DataLoader with background data upload." to "Wraps an existing PyTorch DataLoader with background data upload and allocates each batch to different devices" or something similar would help future users.

**Comment by taylanbil** [comment]
Parallel loader does not change anything about the dataloader's behavior. All it does is send batches to the device in the background. The behavior you are describing is controlled by the `sampler` argument, as seen [here](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py#L165). If this sampler was set to `None`, then every device would receive the same batch.

**Comment by hyang0129** [comment]
Yep its kinda not obvious but logically straight forward.
